### PR TITLE
Use index for peer recovery instead of translog

### PR DESCRIPTION
**Pull request by DaveCTurner**
Today we recover a replica by copying operations from the primary's translog.
However we also retain some historical operations in the index itself, as long
as soft-deletes are enabled. This commit adjusts peer recovery to use the
operations in the index for recovery rather than those in the translog, and
ensures that the replication group retains enough history for use in peer
recovery by means of retention leases.

Reverts #38904 and #42211
Relates #41536

### CODE REVIEW

**Review by ywelsch — APPROVED** [review]
LGTM

File: server/src/main/java/org/elasticsearch/rest/action/admin/cluster/RestClusterHealthAction.java.bak [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+import org.elasticsearch.rest.action.RestStatusToXContentListener;
+
+import java.io.IOException;
+import java.util.Collections;
+import java.util.Locale;
+import java.util.Set;
+
+import static org.elasticsearch.client.Requests.clusterHealthRequest;
+
+public class RestClusterHealthAction extends BaseRestHandler {
```
Comment (dnhatn) [inline server/src/main/java/org/elasticsearch/rest/action/admin/cluster/RestClusterHealthAction.java.bak]: Can you remove this file?

**Review by dnhatn — APPROVED** [review]
LGTM

File: server/src/main/java/org/elasticsearch/rest/action/admin/cluster/RestClusterHealthAction.java.bak (hunk 1 above)
Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/rest/action/admin/cluster/RestClusterHealthAction.java.bak]: Of course, well spotted.

**Comment by DaveCTurner** [comment]
@elasticmachine please run elasticsearch-ci/2

**Pull request merged**